src/cargo/core/resolver/features.rs

```diff
@@ -717,7 +717,14 @@ impl<'a, 'cfg> FeatureResolver<'a, 'cfg> {
                         // The old behavior before weak dependencies were
                         // added is to also enables a feature of the same
                         // name.
-                        self.activate_rec(pkg_id, fk, dep_name)?;
+                        //
+                        // Don't enable if the implicit optional dependency
+                        // feature wasn't created due to `dep:` hiding.
```

Comment:
Feel like it's a bit too concise. Perhaps add some links as references?

```suggestion
                        // Don't enable if the implicit optional dependency
                        // feature wasn't created due to `dep:` hiding.
                        // See rust-lang/cargo#10788 and rust-lang/cargo#12130
```